1)Замена реализации (TreeSet и HashSet) на интерфейс Set - удобнее в …

AlksSrg opened the pull request:
Добавка фичи по **Функциональность «Рекомендации»** + доработка кода
1) Замена реализации (TreeSet и HashSet) на интерфейс Set - удобно при дальнейшем развитии и изменении проекта - не надо переписывать пол кода в случае изменений :
- GenreDao (аргументы методов addGenres и updateGenres);
- GenreDaoImpl (аргументы методов addGenres и updateGenres);
- Film(поле private Set<Genre> genres). 
2) Для реализации фичи добавил в FilmDbService метод по получению фильмов, которые лайкал пользователь. 
3) Для реализации в класс FilmDbStorage метод с sql запросом к БД. 
4) Класс - интерфейс - добавлен метод getFilmsByUser для получения коллекции фильмов которые понравились пользователю. 
5) InMemoryFilmStorage - добавлен метод "заглушка" для избежания ошибки при имплементации интерфейса FilmStorage.

AlksSrg committed:
Добавка фичи по Функциональность «Рекомендации» + доработка кода
1) Замена реализации (TreeSet и HashSet) на интерфейс Set - удобно при дальнейшем развитии и изменении проекта - не надо переписывать пол кода в случае изменений:
- GenreDao (аргументы методов addGenres и updateGenres);
- GenreDaoImpl (аргументы методов addGenres и updateGenres);
- Film(поле private Set<Genre> genres).
2) Для реализации фичи добавил в FilmDbService метод по получению фильмов, которые лайкал пользователь.
3) Для реализации в класс FilmDbStorage метод с sql запросом к БД.
4) Класс - интерфейс - добавлен метод getFilmsByUser для получения коллекции фильмов которые понравились пользователю.
5) InMemoryFilmStorage - добавлен метод "заглушка" для избежания ошибки при имплементации интерфейса FilmStorage.

AlksSrg committed:
1)Замена реализации (TreeSet и HashSet) на интерфейс Set - удобнее в дальнейшей реализации и изменении проекта - не надо переписывать пол кода в случае изменений в дальнейшем:
- GenreDao (аргументы методов addGenres и updateGenres);
- GenreDaoImpl (аргументы методов addGenres и updateGenres);
- Film(поле private Set<Genre> genres).
2)Для реализации фичи добавил в FilmDbService метод по получению фильмов, которые лайкал пользователь.
3)Для реализации в класс FilmDbStorage метод с sql запросом к БД.
4)Класс - интерфейс - добавлен метод getFilmsByUser для получения коллекции фильмов которые понравились пользователю.
5)InMemoryFilmStorage - добавлен метод "заглушка" для избежания ошибки при имплементации интерфейса FilmStorage.

AlksSrg committed:
Нельзя сидеть ночами за кодом=)
Всем хорошего дня =)))

AlksSrg committed:
Merge remote-tracking branch 'origin/add-recommendations' into add-recommendations

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java
#	src/main/java/ru/yandex/practicum/filmorate/service/UserDbService.java

AlksSrg committed:
Нельзя сидеть ночами за кодом=)
Всем хорошего дня =)))

AlksSrg committed:
Нельзя сидеть ночами за кодом=)V3.0 (Чек стайл)
Всем хорошего дня =)))

AlksSrg committed:
Merge remote-tracking branch 'origin/add-recommendations' into add-recommendations

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/storage/films/FilmDbStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/films/FilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/films/InMemoryFilmStorage.java

Irina-Belikova commented on src/main/java/ru/yandex/practicum/filmorate/service/UserDbService.java:
```diff
@@ -182,4 +187,42 @@ public Collection<User> getAllUsers() {
+     * @return возвращает список рекомендуемых фильмов или пустой список если таргет недостаточно обогащен.
+     * @throws EntityNotFoundException генерирует ошибку в случае если пользователь не зарегистрирован.
+     */
+    public List<Film> getRecommendations(long id) {
```
меня смущает история с n+1 запросом к БД

Irina-Belikova commented:
В общем, во всех методах, где внутри цикла есть обращение к БД смущает наличие n+1 запроса к ней.

AlksSrg committed:
Доработка фичи по Рекомендациям:
- увеличиение скорости работы с БД путем получения всех данных в память одним запросом и дальнейшая работа уже с данными из памяти

AlksSrg committed:
Merge branch 'develop' into add-recommendations

AlksSrg committed:
Правки после мержи
